### PR TITLE
Starks Network grant application

### PR DESCRIPTION
**Pull request by xz-cn**
# Grant Application Checklist

- [x] The [application-template.md](https://github.com/w3f/Small-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( "project_name.md") and updated.
- [x] A BTC address for the payment of the milestones is provided inside the application.  
- [x] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19).
- [x] The total funding amount of the project is below 30k at the time of submission.

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application via the open grants program and splitting up the project. I have few additional questions: 

- First would it be possible to reduce the price a little bit further. Your general grants application was a little bit cheaper. 
- Second are you in contact with the original developer of distaff (https://github.com/GuildOfWeavers/distaff). It would be nice to involve him or at least he should be aware of it. 
- Also do you have any previous work around STARK/SNARK or substrate related, which you can share.

**Comment by xz-cn** [comment]
> Thanks for the application via the open grants program and splitting up the project. I have few additional questions:
> 
> * First would it be possible to reduce the price a little bit further. Your general grants application was a little bit cheaper.
> * Second are you in contact with the original developer of distaff (https://github.com/GuildOfWeavers/distaff). It would be nice to involve him or at least he should be aware of it.
> * Also do you have any previous work around STARK/SNARK or substrate related, which you can share.

Hello! Thank you for your comments. Regarding your questions:
1. We have adjusted the amount of BTC according to the latest market price ($10928).
2. We had exchanged tens of emails with the original developer of the Distaff VM to discuss this project. He had also reviewed the full proposal for us. We will involve him in the project somehow in the future but not right now (for his private reasons).
3. We had developed substrate-based projects before, such as a permissioned blockchain for credential (with the help of KILT) and a dex blockchain project. Both were non-open source project though. In addition, we have good connection with the technical community of Polkadot and some parity employees in China who are willing to help with our technical problems.

Please feel free to ask if you have any other questions. Thank you!

**Review by Noc2 — APPROVED** [review]
Thanks for all the additional information, also via email. I’m happy to support the project and will ask the rest of the committee to take a look at the proposal.

**Comment by xz-cn** [comment]
> Thanks for all the additional information, also via email. I’m happy to support the project and will ask the rest of the committee to take a look at the proposal.

Thank you very much for your support!

**Comment by mmagician** [comment]
@xz-cn In the section `Architecture of the Starks Network` you talk about Distaff VM embedded into a *smart contract*, whereas later you mention that it would be a module included in the Substrate Runtime.
So do I understand correctly that you're completely moving away from smart contracts and placing **all** the logic in a pallet, or is there still a need for the smart contract?

Apart from that, could I please ask you to specify at least what public dispatchable functions would be available as part of your pallet? If you already have an idea of other high-level module specifications (e.g. Storage, Events etc.) that would also be very welcome :)

Otherwise the application looks good, I'm looking forward to seeing it in action! Thanks

**Comment by xz-cn** [comment]
> @xz-cn In the section `Architecture of the Starks Network` you talk about Distaff VM embedded into a _smart contract_, whereas later you mention that it would be a module included in the Substrate Runtime.
> So do I understand correctly that you're completely moving away from smart contracts and placing **all** the logic in a pallet, or is there still a need for the smart contract?
> 
> Apart from that, could I please ask you to specify at least what public dispatchable functions would be available as part of your pallet? If you already have an idea of other high-level module specifications (e.g. Storage, Events etc.) that would also be very welcome :)
> 
> Otherwise the application looks good, I'm looking forward to seeing it in action! Thanks

Hello @mmagician , thank you for your comments. Regarding your questions:

1. In the architecture diagram of the Starks Network, the part that was embedded into a _smart contract_ is actually a _Distaff Program_ instead of the Distaff VM. A Distaff Program is constructed using the Distaff assembly to represent the computation which the owner of the smart contract wants the users to perform. We are just using the smart contract as a vehicle for passing the Distaff Program to the users. But it is also ok to pass it using some off-chain method. The Distaff VM itself will indeed be made as part of the native runtime of the Substrate node. The idea is that it can be accessed by an off-chain worker to perform proof verifications via the `runtime_interface`. 

2. The sole function of the Distaff VM runtime module is to perform a computation for zk-STARK proof verification. So it is unnecessary for it to have a storage function. The signature of an example dispatchable function may look like this:
`	fn system_distaff(&self, program_hash: Bytes, public_inputs: Bytes, outputs : Bytes, proof : Bytes) -> SystemResult<String>;
`
Basically, the off-chain worker will provide some hash info and the proof data for the VM to generate a pass/fail result. An event indicating this result and maybe some warning events in case of mistaken input data will be designed. 

I hope this answers your questions. And please feel free to asking if any part is still unclear.

**Comment by mmagician** [comment]
@xz-cn Thanks, that makes sense. I've approved your proposal.

**Review by mmagician — APPROVED** [review]

**Comment by xz-cn** [comment]
@mmagician Thank you for your support!

**Review by RouvenP — APPROVED** [review]

**Comment by xz-cn** [comment]
Thank you to all of you for your support of our project!
